core/deployment/src/main/java/io/quarkus/deployment/IsDockerWorking.java

```diff
@@ -20,90 +23,163 @@
+                        LOGGER.warnf(
+                                "Unable to connect to DOCKER_HOST URI %s, make sure docker is running on the specified host",
+                                dockerHost);
+                    }
+                } catch (URISyntaxException e) {
+                    LOGGER.warnf("Unable to parse DOCKER_HOST URI %s, it will be ignored for working docker detection",
+                            dockerHost);
+                }
+            }
+            return Result.UNKNOWN;
+        }
+    }
+
+    private static class DockerBinaryStrategy implements Strategy {
+
+        private final boolean silent;
+
+        private DockerBinaryStrategy(boolean silent) {
+            this.silent = silent;
+        }
 
-                    for (String line = reader.readLine(); line != null; line = reader.readLine()) {
-                        builder.append(line);
+        @Override
+        public Result get() {
+            try {
+                if (!ExecUtil.execSilent("docker", "-v")) {
+                    LOGGER.warn("'docker -v' returned an error code. Make sure your Docker binary is correct");
+                    return Result.UNKNOWN;
+                }
+            } catch (Exception e) {
+                LOGGER.warnf("No Docker binary found or general error: %s", e);
+                return Result.UNKNOWN;
+            }
+
+            try {
+                OutputFilter filter = new OutputFilter();
+                if (ExecUtil.exec(new File("."), filter, "docker", "version", "--format", "'{{.Server.Version}}'")) {
+                    LOGGER.debugf("Docker daemon found. Version: %s", filter.getOutput());
+                    return Result.AVAILABLE;
+                } else {
+                    if (!silent) {
+                        LOGGER.warn("Could not determine version of Docker daemon");
                     }
-                } catch (IOException e) {
-                    throw new RuntimeException("Error reading stream.", e);
+                    return Result.UNAVAILABLE;
                 }
-            };
+            } catch (Exception e) {
+                LOGGER.warn("Unexpected error occurred while determining Docker daemon version", e);
+                return Result.UNKNOWN;
+            }
         }
 
-        public String getOutput() {
-            return builder.toString();
+        public static class OutputFilter implements Function<InputStream, Runnable> {
+            private final StringBuilder builder = new StringBuilder();
+
+            @Override
+            public Runnable apply(InputStream is) {
+                return () -> {
+                    try (InputStreamReader isr = new InputStreamReader(is);
+                            BufferedReader reader = new BufferedReader(isr)) {
+
+                        for (String line = reader.readLine(); line != null; line = reader.readLine()) {
+                            builder.append(line);
+                        }
```

Comment:
This can also be simplified to:

```suggestion
                  output = new String(is.readFully());
```

Reply:
Good point!

Reply:
Actually, it's not exactly the same thing, so I'll leave it as is

Reply:
Basically there is subtly difference of using the reader, so let's keep it as is

Reply:
Oh? I thought it wouldn't matter because the full output is used in the end, but that's okay 👍🏻